thread.c

```diff
@@ -2639,76 +2636,72 @@ rb_ec_reset_raised(rb_execution_context_t *ec)
     return 1;
 }
 
-int
-rb_notify_fd_close(int fd, struct rb_io_close_wait_list *busy)
+static VALUE
+thread_io_close_notify_all(VALUE _io)
 {
-    rb_vm_t *vm = GET_THREAD()->vm;
-    struct waiting_fd *wfd = 0, *next;
-    ccan_list_head_init(&busy->pending_fd_users);
-    int has_any;
-    VALUE wakeup_mutex;
+    struct rb_io *io = (void*)_io;
 
-    RB_VM_LOCK_ENTER();
-    {
-        ccan_list_for_each_safe(&vm->waiting_fds, wfd, next, wfd_node) {
-            if (wfd->fd == fd) {
-                rb_thread_t *th = wfd->th;
-                VALUE err;
-
-                ccan_list_del(&wfd->wfd_node);
-                ccan_list_add(&busy->pending_fd_users, &wfd->wfd_node);
-
-                wfd->busy = busy;
-                err = th->vm->special_exceptions[ruby_error_stream_closed];
-                rb_threadptr_pending_interrupt_enque(th, err);
-                rb_threadptr_interrupt(th);
-            }
-        }
-    }
+    size_t count = 0;
+    rb_vm_t *vm = io->closing_ec->thread_ptr->vm;
+    VALUE error = vm->special_exceptions[ruby_error_stream_closed];
+
+    struct rb_io_blocking_operation *blocking_operation;
+    ccan_list_for_each(&io->blocking_operations, blocking_operation, list) {
+        rb_execution_context_t *ec = blocking_operation->ec;
+
+        rb_thread_t *thread = ec->thread_ptr;
+        rb_threadptr_pending_interrupt_enque(thread, error);
+
+        // This operation is slow:
+        rb_threadptr_interrupt(thread);
 
-    has_any = !ccan_list_empty(&busy->pending_fd_users);
-    busy->closing_thread = rb_thread_current();
-    busy->closing_fiber = rb_fiber_current();
-    wakeup_mutex = Qnil;
-    if (has_any) {
-        wakeup_mutex = rb_mutex_new();
-        RBASIC_CLEAR_CLASS(wakeup_mutex); /* hide from ObjectSpace */
+        count += 1;
     }
-    busy->wakeup_mutex = wakeup_mutex;
 
-    RB_VM_LOCK_LEAVE();
+    return (VALUE)count;
+}
 
-    /* If the caller didn't pass *busy as a pointer to something on the stack,
-       we need to guard this mutex object on _our_ C stack for the duration
-       of this function. */
-    RB_GC_GUARD(wakeup_mutex);
-    return has_any;
+size_t
+rb_thread_io_close(struct rb_io *io)
+{
+    // We guard this operation based on `io->closing_ec` -> only one thread will ever enter this function.
+    if (io->closing_ec) return 0;
+
+    // Otherwise, we are now closing the IO:
+    rb_execution_context_t *ec = GET_EC();
+    io->closing_ec = ec;
+
+    // If there are no blocking operations, we are done:
+    if (ccan_list_empty(&io->blocking_operations)) return 0;
+
+    // This is used to ensure the correct execution context is woken up after the blocking operation is interrupted:
+    io->wakeup_mutex = rb_mutex_new();
+
+    size_t count = (size_t)rb_mutex_synchronize(io->wakeup_mutex, thread_io_close_notify_all, (VALUE)io);
```

Comment:
Mutex is needed?
At a glance, `thread_io_close_notify_all` seems to have no thread switching points.

Reply:
See https://github.com/ruby/ruby/pull/7884 for why it was introduced. It is also used for compatibility with the fiber scheduler, but maybe we can re-evaluate and consider removal.

Reply:
Okay, I've replaced the initial mutex usage with `RUBY_ASSERT_CRITICAL_SECTION_ENTER`/`RUBY_ASSERT_CRITICAL_SECTION_LEAVE` so we don't accidentally break this code.